pull fails with exception when there is too many repositories in the workspace

```
Searching content for 'filename:gradle.properties repo:manyreposhere'!
org.kohsuke.github.GHException: Failed to retrieve https://api.github.com/search/code?q=org%3Aagorapulse+filename%3Agradle.properties+repo:manyreposhere
	at org.kohsuke.github.GitHubPageIterator.fetch(GitHubPageIterator.java:155)
	at org.kohsuke.github.GitHubPageIterator.hasNext(GitHubPageIterator.java:91)
	at org.kohsuke.github.PagedSearchIterable$1.hasNext(PagedSearchIterable.java:86)
	at org.kohsuke.github.PagedIterator.fetch(PagedIterator.java:106)
	at org.kohsuke.github.PagedIterator.hasNext(PagedIterator.java:74)
	at java.util.Spliterators$IteratorSpliterator.tryAdvance(Spliterators.java:1811)
	at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
	at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at com.agorapulse.pierrot.cli.mixin.SearchMixin.searchContent(SearchMixin.java:81)
	at com.agorapulse.pierrot.cli.PullCommand.run(PullCommand.java:61)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1939)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2346)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2311)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at com.agorapulse.pierrot.cli.PierrotCommand.main(PierrotCommand.java:93)
Caused by: org.kohsuke.github.HttpException: {"message":"Validation Failed","errors":[{"message":"The search is longer than 500 characters.","resource":"Search","field":"q","code":"invalid"}],"documentation_url":"https://docs.github.com/v3/search/"}
	at org.kohsuke.github.GitHubClient.interpretApiError(GitHubClient.java:483)
	at org.kohsuke.github.GitHubClient.sendRequest(GitHubClient.java:407)
	at org.kohsuke.github.GitHubPageIterator.fetch(GitHubPageIterator.java:144)
	... 23 more
Caused by: java.io.IOException: Server returned HTTP response code: 422 for URL: https://api.github.com/search/code?q=org%3Aagorapulse+filename%3Agradle.properties+repo%3Aagorapulse%2Fmicronaut-facebook-sdk%2Cagorapulse%2Fmicronaut-aws-sdk%2Cagorapulse%2Fmicronaut-ap-security%2Cagorapulse%2Fmicronaut-segment%2Cagorapulse%2Fidentity%2Cagorapulse%2Fgroovy-aws-sdk-ses%2Cagorapulse%2Flazybones-templates%2Cagorapulse%2Fgrails-facebook-sdk%2Cagorapulse%2Fnotification%2Cagorapulse%2Fci%2Cagorapulse%2Fmicronaut-snitch%2Cagorapulse%2Fplatform%2Cagorapulse%2Fmicronaut-grails%2Cagorapulse%2Fbusiness%2Cagorapulse%2Ftesting-libraries%2Cagorapulse%2Fmicronaut-bigquery%2Cagorapulse%2Fgroovy-aws-sdk-s3%2Cagorapulse%2Fgrails-segment%2Cagorapulse%2Fdru%2Cagorapulse%2Fgrails-mn-issue%2Cagorapulse%2Fmicronaut-aws-api-gateway-proxy-starter%2Cagorapulse%2Fgroovy-code-checks%2Cagorapulse%2Fgrails-sentry%2Cagorapulse%2Flive%2Cagorapulse%2Fmicronaut-redis-toolbox%2Cagorapulse%2Fgrails-ap-worker%2Cagorapulse%2Fgru%2Cagorapulse%2Fmicronaut-newrelic%2Cagorapulse%2Fmicronaut-recurly%2Cagorapulse%2Fremember%2Cagorapulse%2Fmicronaut-worker%2Cagorapulse%2Fpierrot%2Cagorapulse%2Fvertx-ap-live%2Cagorapulse%2Fmfg-test%2Cagorapulse%2Fmicronaut-rethrow
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1924)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1520)
	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:527)
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:334)
	at org.kohsuke.github.GitHubHttpUrlConnectionClient.getResponseInfo(GitHubHttpUrlConnectionClient.java:64)
	at org.kohsuke.github.GitHubClient.sendRequest(GitHubClient.java:387)
	... 24 more
```

--- a/apps/pierrot/src/main/java/com/agorapulse/pierrot/cli/PierrotCommand.java
+++ b/apps/pierrot/src/main/java/com/agorapulse/pierrot/cli/PierrotCommand.java
@@ -67,6 +67,10 @@ public class PierrotCommand implements Runnable {
     String token;
 
     public static void main(String[] args) {
+        System.exit(execute(args));
+    }
+
+    static int execute(String[] args) {
         if (args.length == 0) {
             args = new String[] {"--help"};
         }
@@ -92,8 +96,7 @@ public class PierrotCommand implements Runnable {
             CommandLine cmd = new CommandLine(PierrotCommand.class, new MicronautFactory(ctx));
             exitCode = cmd.execute(args);
         }
-
-        System.exit(exitCode);
+        return exitCode;
     }
 
     public void run() {

--- a/apps/pierrot/src/main/java/com/agorapulse/pierrot/cli/PullCommand.java
+++ b/apps/pierrot/src/main/java/com/agorapulse/pierrot/cli/PullCommand.java
@@ -18,6 +18,7 @@
 package com.agorapulse.pierrot.cli;
 
 import com.agorapulse.pierrot.api.GitHubService;
+import com.agorapulse.pierrot.api.Repository;
 import com.agorapulse.pierrot.api.ws.Workspace;
 import com.agorapulse.pierrot.cli.mixin.SearchMixin;
 import com.agorapulse.pierrot.cli.mixin.WorkspaceMixin;
@@ -51,17 +52,42 @@ public class PullCommand implements Runnable {
 
     @Override
     public void run() {
+        List<String> repositoryFullNames = new ArrayList<>();
         if (workspaceRepositoriesOnly) {
             Workspace w = new Workspace(workspace.getWorkspace());
-            List<String> repositoryFullNames = new ArrayList<>();
             w.visitRepositories(r -> repositoryFullNames.add(r.getName()));
-            search.constraintToRepositories(repositoryFullNames);
         }
 
         search.searchContent(service, content -> {
-            File location = new File(workspace.getWorkspace(), String.format("%s/%s", content.getRepository().getFullName(), content.getPath()));
+            String repositoryFullName = content.getRepository().getFullName();
+
+            if (!repositoryFullNames.isEmpty() && !repositoryFullNames.contains(repositoryFullName)) {
+                System.out.printf("Repository %s is present in the current workspace.%n", repositoryFullName);
+                return Optional.empty();
+            }
+
+            Optional<Repository> maybeRepository = service.getRepository(repositoryFullName);
+
+            if (maybeRepository.isEmpty()) {
+                System.out.printf("Repository %s is not available.%n", repositoryFullName);
+                return Optional.empty();
+            }
+
+            Repository ghr = maybeRepository.get();
+
+            if (ghr.isArchived()) {
+                System.out.printf("Repository %s is archived.%n", ghr.getFullName());
+                return Optional.empty();
+            }
+
+            if (!ghr.canWrite()) {
+                System.out.printf("Current user does not have write rights to the repository %s.%n", ghr.getFullName());
+                return Optional.empty();
+            }
+
+            File location = new File(workspace.getWorkspace(), String.format("%s/%s", repositoryFullName, content.getPath()));
             content.writeTo(location);
-            System.out.printf("Fetched %s/%s%n", content.getRepository().getFullName(), content.getPath());
+            System.out.printf("Fetched %s/%s%n", repositoryFullName, content.getPath());
             return Optional.of(location.toURI());
         });
 

--- a/apps/pierrot/src/main/java/com/agorapulse/pierrot/cli/PullCommand.java
+++ b/apps/pierrot/src/main/java/com/agorapulse/pierrot/cli/PullCommand.java
@@ -62,7 +62,7 @@ public class PullCommand implements Runnable {
             String repositoryFullName = content.getRepository().getFullName();
 
             if (!repositoryFullNames.isEmpty() && !repositoryFullNames.contains(repositoryFullName)) {
-                System.out.printf("Repository %s is present in the current workspace.%n", repositoryFullName);
+                System.out.printf("Ignoring %s/%s as the repository is not pulled%n", repositoryFullName, content.getPath());
                 return Optional.empty();
             }
 

--- a/apps/pierrot/src/main/java/com/agorapulse/pierrot/cli/mixin/SearchMixin.java
+++ b/apps/pierrot/src/main/java/com/agorapulse/pierrot/cli/mixin/SearchMixin.java
@@ -134,12 +134,6 @@ public class SearchMixin {
         }
     }
 
-    public void constraintToRepositories(List<String> repositoryFullNames) {
-        if (!repositoryFullNames.isEmpty()) {
-            queries.add("repo:" + String.join(",", repositoryFullNames));
-        }
-    }
-
     private boolean isNoPage() {
         if (noPage) {
             return true;

--- a/apps/pierrot/src/test/groovy/com/agorapulse/pierrot/cli/AbstractCommandSpec.groovy
+++ b/apps/pierrot/src/test/groovy/com/agorapulse/pierrot/cli/AbstractCommandSpec.groovy
@@ -169,10 +169,6 @@ abstract class AbstractCommandSpec extends Specification {
             Stream.of(content1, content2, content3)
         }
 
-        searchContent("$CONTENT_SEARCH_TERM repo:$REPOSITORY_ONE", false) >> {
-            Stream.of(content1, content2, content3)
-        }
-
         searchPullRequests(PR_SEARCH_TERM, true, false) >> {
             Stream.of(pullRequest1, pullRequest2)
         }

--- a/apps/pierrot/src/test/groovy/com/agorapulse/pierrot/cli/PierrotCommandSpec.groovy
+++ b/apps/pierrot/src/test/groovy/com/agorapulse/pierrot/cli/PierrotCommandSpec.groovy
@@ -31,7 +31,7 @@ class PierrotCommandSpec extends AbstractCommandSpec {
     void 'run main'() {
         when:
             TestConsole console = TestConsole.capture {
-                PierrotCommand.main('--help', '--github-token=token', '-s')
+                PierrotCommand.execute('--help', '--github-token=token', '-s')
             }
 
         then:
@@ -41,7 +41,7 @@ class PierrotCommandSpec extends AbstractCommandSpec {
     void 'run main without token'() {
         when:
             TestConsole console = TestConsole.capture {
-                PierrotCommand.main('--help')
+                PierrotCommand.execute('--help')
             }
 
         then:

--- a/apps/pierrot/src/test/groovy/com/agorapulse/pierrot/cli/PullCommandSpec.groovy
+++ b/apps/pierrot/src/test/groovy/com/agorapulse/pierrot/cli/PullCommandSpec.groovy
@@ -44,10 +44,8 @@ class PullCommandSpec extends AbstractCommandSpec {
 
         assert file1.exists()
         assert file1.text == CONTENT
-        assert file2.exists()
-        assert file2.text == CONTENT.reverse()
-        assert file3.exists()
-        assert file3.text == CONTENT
+        assert !file2.exists()
+        assert !file3.exists()
 
         return true
     }

--- a/apps/pierrot/src/test/resources/com/agorapulse/pierrot/cli/PullCommandSpec/run.txt
+++ b/apps/pierrot/src/test/resources/com/agorapulse/pierrot/cli/PullCommandSpec/run.txt
@@ -1,5 +1,5 @@
-Searching content for 'org:agorapulse filename:.testfile repo:agorapulse/pierrot'!
+Searching content for 'org:agorapulse filename:.testfile'!
 Fetched agorapulse/pierrot/.testfile
-Fetched agorapulse/oss/.testfile
-Fetched agorapulse/oss/prefix/.testfile
+Ignoring agorapulse/oss/.testfile as the repository is not pulled
+Ignoring agorapulse/oss/prefix/.testfile as the repository is not pulled
 Found 3 results!

